src/subdomains/core/custody/services/custody.service.ts

```diff
@@ -0,0 +1,171 @@
+import { CustodyAssetBalanceDtoMapper } from '../mappers/custody-asset-balance-dto.mapper';
+import { CustodyBalanceRepository } from '../repositories/custody-balance.repository';
+import { CustodyOrderStepRepository } from '../repositories/custody-order-step.repository';
+import { CustodyOrderRepository } from '../repositories/custody-order.repository';
+
+@Injectable()
+export class CustodyService {
+  constructor(
+    private readonly userService: UserService,
+    private readonly userDataService: UserDataService,
+    private readonly walletService: WalletService,
+    private readonly refService: RefService,
+    private readonly authService: AuthService,
+    private readonly custodyOrderRepo: CustodyOrderRepository,
+    private readonly custodyOrderStepRepo: CustodyOrderStepRepository,
+    private readonly custodyBalanceRepo: CustodyBalanceRepository,
+  ) {}
+
+  //*** PUBLIC API ***//
+  async createCustodyAccount(accountId: number, dto: CreateCustodyAccountDto, userIp: string): Promise<CustodyAuthDto> {
+    const ref = await this.refService.get(userIp);
+    if (ref) dto.usedRef ??= ref.ref;
+
+    const wallet =
+      (await this.walletService.getByIdOrName(undefined, dto.wallet)) ?? (await this.walletService.getDefault());
+
+    const addressIndex = await this.userService.getNexCustodyIndex();
+    const custodyWallet = EvmUtil.createWallet(Config.blockchain.evm.custodyAccount(addressIndex));
+    const signature = await custodyWallet.signMessage(Config.auth.signMessageGeneral + custodyWallet.address);
+
+    const account = await this.userDataService.getUserData(accountId);
+    if (!account) throw new NotFoundException('User not found');
+
+    const custodyUser = await this.userService.createUser(
+      {
+        address: custodyWallet.address,
+        signature,
+        usedRef: dto.usedRef,
+        ip: userIp,
+        origin: ref?.origin,
+        wallet,
+        userData: account,
+        custodyAddressType: dto.addressType,
+        custodyAddressIndex: addressIndex,
+        role: UserRole.CUSTODY,
+      },
+      dto.specialCode,
+    );
+
+    return { accessToken: this.authService.generateUserToken(custodyUser, userIp) };
+  }
+
+  async createStep(
+    order: CustodyOrder,
+    index: number,
+    command: string,
+    context: CustodyOrderStepContext,
+  ): Promise<CustodyOrderStep> {
+    const orderStep = this.custodyOrderStepRepo.create({
+      order,
+      index,
+      command,
+      context,
+    });
+    return this.custodyOrderStepRepo.save(orderStep);
+  }
+
+  async startNextStep(step: CustodyOrderStep): Promise<void> {
+    const nextIndex = step.index + 1;
+    const order = step.order;
+    const nextStep = OrderConfig[order.type][nextIndex];
+
+    if (nextStep) {
+      await this.createStep(order, nextIndex, nextStep.command, nextStep.context);
+    } else {
+      const asset = order.inputAsset ?? order.outputAsset;
+      const amount = order.inputAmount ?? -order.outputAmount;
```

Comment:
There might be orders with input and output asset/amount -> we need to update both balances?